cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -309,6 +314,240 @@ Result<std::vector<FileInfo>> LocalFileSystem::GetFileInfo(const FileSelector& s
   return results;
 }
 
+namespace {
+
+/// Workhorse for streaming async implementation of `GetFileInfo`
+/// (`GetFileInfoGenerator`).
+///
+/// There are two variants of async discovery functions suported:
+/// 1. `DiscoverDirectoryFiles`, which parallelizes traversal of individual directories
+///    so that each directory results are yielded as a separate `FileInfoGenerator` via
+///    an underlying `DiscoveryImplIterator`, which delivers items in chunks (default size
+///    is 1K items).
+/// 2. `DiscoverDirectoriesFlattened`, which forwards execution to the
+///    `DiscoverDirectoryFiles`, with the difference that the results from individual
+///    sub-directory iterators are merged into the single FileInfoGenerator stream.
+///
+/// The implementation makes use of additional attributes in `LocalFileSystemOptions`,
+/// such as `directory_readahead`, which can be used to tune algorithm
+/// behavior and adjust how many directories can be processed in parallel.
+/// This option is disabled by default, so that individual directories are processed
+/// in serial manner via `MakeConcatenatedGenerator` under the hood.
+class AsyncStatSelector {
+ public:
+  using FileInfoGeneratorProducer = PushGenerator<FileInfoGenerator>::Producer;
+
+  /// Helper class, which wraps the producer instance and ensures
+  /// that it stays "open" (i.e. `producer.is_closed() == false`)
+  /// until all discovery iterators are exhausted.
+  ///
+  /// The calss maintains a shared `State`, which wraps the original producer
+  /// and automatically calls `Close()` on it once the ref-count for the
+  /// state reaches zero (which is equivalent to finishing the file discovery
+  /// process).
+  class AutoClosingProducer {
```

Comment:
`PushGenerator` is thread-safe, but unfortunately `AutoClosingProducer` is not.

It seems you can vastly simplify this and make it thread-safe by letting C++ do the work:
```c++
struct DiscoveryState {
  FileInfoGeneratorProducer producer;

  ~DiscoveryState() {
    producer->Close();
  }
};
```
... then pass the same `std::shared_ptr<DiscoveryState>` to every `DiscoveryImplIterator`:
```c++
  static Result<AsyncGenerator<FileInfoGenerator>> DiscoverDirectoryFiles(
      FileSelector selector, LocalFileSystemOptions fs_opts) {
    PushGenerator<FileInfoGenerator> file_gen;

    ARROW_ASSIGN_OR_RAISE(
        auto base_dir, arrow::internal::PlatformFilename::FromString(selector.base_dir));
    auto discovery_state = std::make_shared<DiscoveryState>(std::move(file_gen.producer()));
    ARROW_RETURN_NOT_OK(DoDiscovery(std::move(base_dir), 0, std::move(selector),
                                    std::move(discovery_state),
                                    fs_opts.file_info_batch_size));

    return file_gen;
  }
```

Reply:
Good point, thanks for catching it!